File: backend/internal/sidecar/middleware.go
package sidecar

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"math"
	"net/http"
	"net/url"
	"os"
	"time"

	"github.com/your-org/stellarflow/internal/x402"

	"github.com/gin-gonic/gin"
)

const (
	headerXPayment       = "X-Payment"
	x402SchemeExact      = "exact"
	x402Version          = 2
	defaultMaxTimeoutSec = 30
)

// X402Config holds the Stellar network + asset config for building 402 challenges.
type X402Config struct {
	Network         string // "stellar:testnet"
	USDCContract    string // Soroban SAC C-address
	USDCIssuer      string // classic G-issuer (informational)
	USDCCode        string // "USDC"
	TreasuryAddress string // G-address that receives payments
	DryRun          bool
}

// X402Middleware returns a gin middleware that gates a request behind an x402
// payment of the given USDC amount. It uses the facilitator to verify + settle,
// and calls the provided onPaid callback with metadata for logging/SSE.
func X402Middleware(
	cfg X402Config,
	facilitator *x402.Client,
	endpoint string,
	priceUSDC float64,
	onPaid func(call PaidCall),
) gin.HandlerFunc {
	return func(ctx *gin.Context) {
		now := time.Now()

		requirement := x402.PaymentRequirements{
			Scheme:            x402SchemeExact,
			Network:           cfg.Network,
			Asset:             cfg.USDCContract,
			Amount:            formatAmount(priceUSDC),
			PayTo:             cfg.TreasuryAddress,
			MaxTimeoutSeconds: defaultMaxTimeoutSec,
			Extra: map[string]interface{}{
				"code":             cfg.USDCCode,
				"issuer":           cfg.USDCIssuer,
				"areFeesSponsored": true,
			},
		}
		paymentRequired := x402.PaymentRequired{
			X402Version: x402Version,
			Resource: x402.ResourceInfo{
				URL:         absoluteURL(ctx),
				Description: endpoint,
				MimeType:    "application/json",
			},
			Accepts: []x402.PaymentRequirements{requirement},
		}

		// No X-PAYMENT header → 402.
		paymentHeader := ctx.GetHeader(headerXPayment)
		if paymentHeader == "" {
			paymentRequired.Error = "X-PAYMENT header missing"
			ctx.JSON(http.StatusPaymentRequired, paymentRequired)
			ctx.Abort()
			return
		}

		// Decode + parse the payload.
		payload, err := decodePaymentPayload(paymentHeader)
		if err != nil {
			paymentRequired.Error = fmt.Sprintf("invalid X-PAYMENT header: %v", err)
			ctx.JSON(http.StatusPaymentRequired, paymentRequired)
			ctx.Abort()
			return
		}

		// Validate the payload matches our requirements.
		if payload.Accepted.Scheme != requirement.Scheme ||
			payload.Accepted.Network != requirement.Network ||
			payload.Accepted.PayTo != requirement.PayTo ||
			payload.Accepted.Amount != requirement.Amount {
			paymentRequired.Error = "payment requirements mismatch"
			ctx.JSON(http.StatusPaymentRequired, paymentRequired)
			ctx.Abort()
			return
		}

		// Verify + settle.
		var facilitatorName string
		var verifiedPayer string
		if cfg.DryRun {
			facilitatorName = "dry-run"
			verifiedPayer = extractDryRunPayer(payload.Payload)
		} else {
			if facilitator == nil {
				paymentRequired.Error = "facilitator not configured"
				ctx.JSON(http.StatusPaymentRequired, paymentRequired)
				ctx.Abort()
				return
			}
			verifyCtx, cancel := context.WithTimeout(ctx.Request.Context(), 15*time.Second)
			defer cancel()
			verifyResp, err := facilitator.Verify(verifyCtx, x402.VerifyRequest{
				X402Version:         x402Version,
				PaymentPayload:      *payload,
				PaymentRequirements: requirement,
			})
			if err != nil {
				paymentRequired.Error = fmt.Sprintf("facilitator verify error: %v", err)
				ctx.JSON(http.StatusPaymentRequired, paymentRequired)
				ctx.Abort()
				return
			}
			if !verifyResp.IsValid {
				paymentRequired.Error = fmt.Sprintf("payment rejected: %s — %s", verifyResp.InvalidReason, verifyResp.InvalidMessage)
				ctx.JSON(http.StatusPaymentRequired, paymentRequired)
				ctx.Abort()
				return
			}
			facilitatorName = "openzeppelin"
			verifiedPayer = verifyResp.Payer
		}

		// Stash payer on context for the proxy handler.
		ctx.Set("payer_address", normalizeStellarAddress(verifiedPayer))

		// Run the proxy (next handler).
		ctx.Next()

		// Post-handler: settle + log.
		go func() {
			defer func() {
				if r := recover(); r != nil {
					fmt.Fprintf(os.Stderr, "[x402-sidecar] PANIC in settle/log: %v\n", r)
				}
			}()

			var txHash string
			if !cfg.DryRun && facilitator != nil {
				bgCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
				defer cancel()
				settleResp, err := facilitator.Settle(bgCtx, x402.VerifyRequest{
					X402Version:         x402Version,
					PaymentPayload:      *payload,
					PaymentRequirements: requirement,
				})
				if err != nil {
					fmt.Fprintf(os.Stderr, "[x402-sidecar] settle error endpoint=%s payer=%s: %v\n", endpoint, verifiedPayer, err)
				} else if !settleResp.Success {
					fmt.Fprintf(os.Stderr, "[x402-sidecar] settle failed endpoint=%s: %s — %s\n", endpoint, settleResp.ErrorReason, settleResp.ErrorMessage)
				} else {
					txHash = settleResp.Transaction
					if settleResp.Payer != "" {
						verifiedPayer = settleResp.Payer
					}
				}
			}

			call := PaidCall{
				Endpoint:     endpoint,
				PayerAddress: normalizeStellarAddress(verifiedPayer),
				PriceUSDC:    priceUSDC,
				Facilitator:  facilitatorName,
				TxHash:       txHash,
				DurationMs:   int(time.Since(now).Milliseconds()),
				Timestamp:    now,
			}
			if onPaid != nil {
				onPaid(call)
			}
		}()
	}
}

// PaidCall is the metadata emitted after a successful paid call.
type PaidCall struct {
	Endpoint     string    `json:"endpoint"`
	PayerAddress string    `json:"payer_address"`
	PriceUSDC    float64   `json:"price_usdc"`
	Facilitator  string    `json:"facilitator"`
	TxHash       string    `json:"tx_hash,omitempty"`
	DurationMs   int       `json:"duration_ms"`
	Timestamp    time.Time `json:"timestamp"`
}

// --- helpers (mirrored from internal/api/middleware_x402.go) ---

func decodePaymentPayload(header string) (*x402.PaymentPayload, error) {
	raw, err := base64.StdEncoding.DecodeString(header)
	if err != nil {
		raw, err = base64.URLEncoding.DecodeString(header)
		if err != nil {
			return nil, fmt.Errorf("base64 decode: %w", err)
		}
	}
	var pl x402.PaymentPayload
	if err := json.Unmarshal(raw, &pl); err != nil {
		return nil, fmt.Errorf("json parse: %w", err)
	}
	if pl.X402Version == 0 {
		pl.X402Version = x402Version
	}
	return &pl, nil
}

func extractDryRunPayer(payload json.RawMessage) string {
	if len(payload) == 0 {
		return ""
	}
	var p struct {
		From string `json:"from"`
	}
	if err := json.Unmarshal(payload, &p); err != nil {
		return ""
	}
	return p.From
}

func normalizeStellarAddress(addr string) string {
	if len(addr) == 56 && addr[0] == 'G' {
		return addr
	}
	return "GAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADRYRUNANONYMOUSXXXXX"
}

func formatAmount(usdc float64) string {
	stroops := int64(math.Round(usdc * 10_000_000))
	return fmt.Sprintf("%d", stroops)
}

func absoluteURL(ctx *gin.Context) string {
	scheme := "http"
	if ctx.Request.TLS != nil || ctx.GetHeader("X-Forwarded-Proto") == "https" {
		scheme = "https"
	}
	host := ctx.GetHeader("X-Forwarded-Host")
	if host == "" {
		host = ctx.Request.Host
	}
	return fmt.Sprintf("%s://%s%s", scheme, host, ctx.Request.URL.Path)
}

// rawReasoningHeader extracts and URL-decodes the reasoning header.
func rawReasoningHeader(ctx *gin.Context) string {
	raw := ctx.GetHeader("X-StellarFlow-Reason")
	if raw == "" {
		return ""
	}
	if dec, err := url.QueryUnescape(raw); err == nil {
		return dec
	}
	return raw
}

File: middleware/python/x402_middleware/middleware.py
"""
Core x402 paywall decorator for FastAPI / Starlette.

No dependencies beyond the Python standard library + httpx for the
facilitator HTTP calls. Works with any ASGI framework that uses
Starlette's Request object (FastAPI, Starlette, etc).
"""

from __future__ import annotations

import base64
import functools
import json
import math
import os
import asyncio
from typing import Any, Callable

try:
    import httpx
except ImportError:
    httpx = None  # type: ignore

# ─── Constants ─────────────────────────────────────────────────

DEFAULT_FACILITATOR_URL = "https://channels.openzeppelin.com/x402"
DEFAULT_NETWORK = "stellar:testnet"
DEFAULT_USDC_CONTRACT = "CBIELTK6YBZJU5UP2WWQEUCYKLPU6AUNZ2BQ4WWFEIE3USCIHMXQDAMA"
DEFAULT_USDC_ISSUER = "GBBD47IF6LWK7P7MDEVSCWR7DPUWV3NY3DTQEVFL4NAT4AQH3ZLLFLA5"
X402_VERSION = 2


def _format_amount(usdc: float) -> str:
    """Convert USDC decimal to integer stroops string (7 decimals)."""
    return str(round(usdc * 10_000_000))


def _normalize_address(addr: str) -> str:
    if len(addr) == 56 and addr.startswith("G"):
        return addr
    return "GAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADRYRUNANONYMOUSXXXXX"


# ─── Decorator factory ─────────────────────────────────────────

def x402_paywall(
    price: float,
    treasury: str | None = None,
    *,
    facilitator_url: str | None = None,
    facilitator_key: str | None = None,
    network: str | None = None,
    usdc_contract: str | None = None,
    usdc_issuer: str | None = None,
    description: str | None = None,
    dry_run: bool | None = None,
) -> Callable:
    """
    Decorator that gates a FastAPI/Starlette endpoint behind an x402 paywall.

    Usage:
        @app.post("/api/sentiment")
        @x402_paywall(price=0.05)  # reads treasury from X402_TREASURY_ADDRESS env
        async def sentiment(request: Request):
            payer = request.state.x402_payer
            return {"result": "..."}
    """
    # Fall back to env vars for everything except price.
    treasury = treasury or os.environ.get("X402_TREASURY_ADDRESS", "")
    facilitator_url = facilitator_url or os.environ.get(
        "X402_FACILITATOR_URL", DEFAULT_FACILITATOR_URL
    )
    facilitator_key = facilitator_key or os.environ.get("X402_FACILITATOR_KEY", "")
    network = network or os.environ.get("X402_NETWORK", DEFAULT_NETWORK)
    usdc_contract = usdc_contract or os.environ.get(
        "X402_USDC_CONTRACT", DEFAULT_USDC_CONTRACT
    )
    usdc_issuer = usdc_issuer or os.environ.get("X402_USDC_ISSUER", DEFAULT_USDC_ISSUER)
    if dry_run is None:
        dry_run = os.environ.get("X402_DRY_RUN", "false").lower() == "true"

    if not treasury:
        raise RuntimeError(
            "x402_paywall: no treasury address. "
            "Set X402_TREASURY_ADDRESS env var or pass treasury= kwarg. "
            "Run `python -m x402_middleware init` to generate one."
        )

    amount_stroops = _format_amount(price)

    requirement = {
        "scheme": "exact",
        "network": network,
        "asset": usdc_contract,
        "amount": amount_stroops,
        "payTo": treasury,
        "maxTimeoutSeconds": 30,
        "extra": {
            "code": "USDC",
            "issuer": usdc_issuer,
            "areFeesSponsored": True,
        },
    }

    def decorator(func: Callable) -> Callable:
        @functools.wraps(func)
        async def wrapper(*args: Any, **kwargs: Any) -> Any:
            # Find the Request object in args/kwargs (FastAPI passes it as 'request').
            request = kwargs.get("request") or (args[0] if args else None)
            if request is None:
                raise RuntimeError(
                    "x402_paywall: could not find Request object. "
                    "Make sure your handler has a 'request: Request' parameter."
                )

            from starlette.requests import Request as StarletteRequest
            from starlette.responses import JSONResponse

            if not isinstance(request, StarletteRequest):
                raise RuntimeError("x402_paywall: first argument must be a Starlette Request")

            url = str(request.url)
            payment_required = {
                "x402Version": X402_VERSION,
                "error": "",
                "resource": {
                    "url": url,
                    "description": description or request.url.path,
                    "mimeType": "application/json",
                },
                "accepts": [requirement],
            }

            # 1. Check for X-PAYMENT header
            payment_header = request.headers.get("x-payment", "")
            if not payment_header:
                payment_required["error"] = "X-PAYMENT header missing"
                return JSONResponse(payment_required, status_code=402)

            # 2. Decode the payload
            try:
                raw = base64.b64decode(payment_header)
                payload = json.loads(raw)
            except Exception:
                payment_required["error"] = "invalid X-PAYMENT header: base64/JSON decode failed"
                return JSONResponse(payment_required, status_code=402)

            # 3. Validate requirements match
            acc = payload.get("accepted", {})
            if (
                acc.get("scheme") != requirement["scheme"]
                or acc.get("network") != requirement["network"]
                or acc.get("payTo") != requirement["payTo"]
                or acc.get("amount") != requirement["amount"]
            ):
                payment_required["error"] = "payment requirements mismatch"
                return JSONResponse(payment_required, status_code=402)

            # 4. Verify with facilitator
            payer = ""
            if dry_run:
                inner = payload.get("payload", {})
                if isinstance(inner, str):
                    try:
                        inner = json.loads(inner)
                    except Exception:
                        inner = {}
                payer = inner.get("from", "GDRYRUNXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX")
            else:
                if not facilitator_key:
                    payment_required["error"] = "facilitator not configured"
                    return JSONResponse(payment_required, status_code=402)

                if httpx is None:
                    raise RuntimeError("x402_paywall requires httpx: pip install httpx")

                async with httpx.AsyncClient(timeout=15) as client:
                    verify_resp = await client.post(
                        f"{facilitator_url.rstrip('/')}/verify",
                        json={
                            "x402Version": X402_VERSION,
                            "paymentPayload": payload,
                            "paymentRequirements": requirement,
                        },
                        headers={
                            "Authorization": f"Bearer {facilitator_key}",
                            "Content-Type": "application/json",
                        },
                    )

                if verify_resp.status_code >= 400:
                    payment_required["error"] = f"facilitator error: {verify_resp.text}"
                    return JSONResponse(payment_required, status_code=402)

                verify_data = verify_resp.json()
                if not verify_data.get("isValid"):
                    reason = verify_data.get("invalidReason", "unknown")
                    payment_required["error"] = f"payment rejected: {reason}"
                    return JSONResponse(payment_required, status_code=402)

                payer = verify_data.get("payer", "")

            # 5. Set payer on request state
            normalized_payer = _normalize_address(payer)
            request.state.x402_payer = normalized_payer
            request.state.x402_amount = price

            import time as _time
            started_at = _time.time()

            # 6. Call the handler
            response = await func(*args, **kwargs)

            # Log to the in-memory dashboard logger (if the handler succeeded).
            if getattr(response, "status_code", 200) < 400:
                from .logger import default_logger, PaidCallLog
                default_logger.log(PaidCallLog(
                    timestamp=started_at,
                    endpoint=description or request.url.path,
                    payer=normalized_payer,
                    amount=price,
                    facilitator="openzeppelin" if not dry_run else "dry-run",
                    duration_ms=int((_time.time() - started_at) * 1000),
                ))

            # 7. Settle async (non-blocking)
            if not dry_run and facilitator_key and httpx is not None:
                asyncio.create_task(_settle(
                    facilitator_url, facilitator_key, payload, requirement
                ))

            return response

        return wrapper
    return decorator


async def _settle(
    facilitator_url: str,
    facilitator_key: str,
    payload: dict,
    requirement: dict,
) -> None:
    """Settle the payment on-chain (fire-and-forget after handler response)."""
    try:
        async with httpx.AsyncClient(timeout=30) as client:
            await client.post(
                f"{facilitator_url.rstrip('/')}/settle",
                json={
                    "x402Version": X402_VERSION,
                    "paymentPayload": payload,
                    "paymentRequirements": requirement,
                },
                headers={
                    "Authorization": f"Bearer {facilitator_key}",
                    "Content-Type": "application/json",
                },
            )
    except Exception as e:
        print(f"[x402-middleware] settle error: {e}")
